src/gui/TrackContentObjectView.cpp

```diff
@@ -1106,6 +1127,122 @@ void TrackContentObjectView::toggleMute( QVector<TrackContentObjectView *> tcovs
 	}
 }
 
+bool TrackContentObjectView::canMergeSelection(QVector<TrackContentObjectView*> tcovs)
+{
+	// We can only merge InstrumentTrack's TCOs, so check if we only have those in the selection,
+	// and also if they all belong to the same track
+
+	// We start assuming we can merge and if any condition is broken we set it to false
+	bool canMerge = true;
+
+	// Variable to check if all TCOs belong to the same track
+	TrackView* previousOwnerTrackView = nullptr;
+
+	// Then we check every selected TCO to see if all of them are InstrumentTrack's TCOs.
+	// If any isn't, we set canMerge to false and quit the loop.
+	for (auto tcov: tcovs)
+	{
+		TrackView* ownerTrackView = tcov->getTrackView();
+
+		// Set the previousOwnerTrackView to the first TrackView
+		if (!previousOwnerTrackView)
+		{
+			previousOwnerTrackView = ownerTrackView;
+		}
+
+		// If there are TCOs from different tracks or TCOs from tracks
+		// other than an InstrumentTrack, can't merge them
+		if (ownerTrackView != previousOwnerTrackView
+			|| !dynamic_cast<InstrumentTrackView*>(ownerTrackView))
+		{
+			canMerge = false;
+			break;
+		}
+	}
+
+	return canMerge;
+}
+
+void TrackContentObjectView::mergeTCOs(QVector<TrackContentObjectView*> tcovs)
+{
+	// Get the track that we are merging TCOs in
+	InstrumentTrack* track =
+		dynamic_cast<InstrumentTrack*>(tcovs.at(0)->getTrackView()->getTrack());
+
+	if (!track)
+	{
+		qWarning("Warning: Couldn't retrieve InstrumentTrack in mergeTCOs()");
+		return;
+	}
+
+	// For Undo/Redo
+	track->addJournalCheckPoint();
+	track->saveJournallingState(false);
+
+	// Find the earliest position of all the selected TCOVs
+	TimePos earliestPos = tcovs.at(0)->getTrackContentObject()->startPosition();
+	TimePos currentPos = earliestPos;
+
+	for (auto tcov: tcovs)
+	{
+		currentPos = tcov->getTrackContentObject()->startPosition();
+		if (currentPos < earliestPos)
+		{
+			earliestPos = currentPos;
+		}
+	}
```

Comment:
Something like this should work?
```suggestion
	const auto earliestTCOV = std::min_element(tcovs.begin(), tcovs.end(),
		[](TrackContentObjectView* a, TrackContentObjectView* b){
			return a->getTrackContentObject()->startPosition() <
			b->getTrackContentObject()->startPosition;
		}
	);
	const TimePos earliestPos = earliestTCOV->getTrackContentObject()->startPosition();
```

Reply:
Are Qt iterators guaranteed to be compatible with methods that expect `std::iterator`s?

Reply:
I think so? https://doc.qt.io/qt-5/containers.html#stl-style-iterators

These can probably be constBegin and constEnd by the way.

Reply:
Cool! I thought they should be, but was asking to make sure. I'm trying to use `std::set` on the other suggestion btw, just to follow that guideline of avoiding Qt where it isn't needed (even though `QVector` is being used for the selection, but that's something for another PR :laughing: )